x/group/keeper/msg_server.go

```diff
@@ -233,6 +233,71 @@ func (k Keeper) UpdateGroupMetadata(goCtx context.Context, req *group.MsgUpdateG
 	return &group.MsgUpdateGroupMetadataResponse{}, nil
 }
 
+func (k Keeper) CreateGroupWithPolicy(goCtx context.Context, req *group.MsgCreateGroupWithPolicy) (*group.MsgCreateGroupWithPolicyResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	groupPolicyMetadata := req.GroupPolicyMetadata
+	if err := assertMetadataLength(groupPolicyMetadata, "group policy metadata"); err != nil {
+		return nil, err
+	}
+
+	groupRes, err := k.CreateGroup(goCtx, &group.MsgCreateGroup{
+		Admin:    req.Admin,
+		Members:  req.Members,
+		Metadata: req.GroupMetadata,
+	})
+	if err != nil {
+		return nil, sdkerrors.Wrap(err, "group response")
+	}
+	groupId := groupRes.GroupId
+
+	var groupPolicyAddr sdk.AccAddress
+	groupPolicyRes, err := k.CreateGroupPolicy(goCtx, &group.MsgCreateGroupPolicy{
+		Admin:          req.Admin,
+		GroupId:        groupId,
+		Metadata:       groupPolicyMetadata,
+		DecisionPolicy: req.DecisionPolicy,
+	})
+	if err != nil {
+		return nil, sdkerrors.Wrap(err, "group policy response")
+	}
+	policyAddr := groupPolicyRes.Address
+
+	groupPolicyAddr, err = sdk.AccAddressFromBech32(policyAddr)
+	if err != nil {
+		return nil, sdkerrors.Wrap(err, "group policy address")
+	}
+
+	if req.GroupPolicyAsAdmin {
+		updateAdminReq := &group.MsgUpdateGroupAdmin{
+			GroupId:  groupId,
+			Admin:    req.Admin,
+			NewAdmin: groupPolicyAddr.String(),
+		}
+		_, err = k.UpdateGroupAdmin(goCtx, updateAdminReq)
+		if err != nil {
+			return nil, err
+		}
+
+		updatePolicyAddressReq := &group.MsgUpdateGroupPolicyAdmin{
+			Admin:    req.Admin,
+			Address:  groupPolicyAddr.String(),
+			NewAdmin: groupPolicyAddr.String(),
+		}
+		_, err = k.UpdateGroupPolicyAdmin(goCtx, updatePolicyAddressReq)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	err = ctx.EventManager().EmitTypedEvents(&group.EventCreateGroup{GroupId: groupId}, &group.EventCreateGroupPolicy{Address: groupPolicyAddr.String()})
+	if err != nil {
+		return nil, err
+	}
```

Comment:
All these events are already emitted in `k.CreateGroup` & `k.CreateGroupPolicy`.